[FEAT] Add new API to remove CSS classes from BPMN elements

tbouffard opened the pull request:
Provide an API close to the one that already exists to add CSS classes.

covers #942

tbouffard commented on src/component/registry/bpmn-elements-registry.ts:
```diff
@@ -112,6 +112,22 @@ export class BpmnElementsRegistry {
+   * @param classNames The name of the class(es) to remove from the BPMN element(s)
+   */
+  removeCssClasses(bpmnElementIds: string | string[], classNames: string | string[]): void {
+    // TODO duplication with addCssClasses
+    // only the call to remove or add class differ
```
ℹ️ Waiting for #1043 to decide when to do the refactoring

tbouffard commented on src/component/registry/css-registry.ts:
```diff
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 export class CssRegistry {
+  // TODO using a Set internally will prevent multi transformation (only when getting the class list)
```
ℹ️ Waiting for #1043 to decide when to do the refactoring

tbouffard replied:
We have decided to use Sets internally

csouchet commented on src/component/registry/bpmn-elements-registry.ts:
```diff
@@ -112,6 +112,22 @@ export class BpmnElementsRegistry {
+    // TODO duplication with addCssClasses
+    // only the call to remove or add class differ
+    ensureIsArray<string>(bpmnElementIds).forEach(bpmnElementId => {
+      if (this.cssRegistry.removeClassNames(bpmnElementId, ensureIsArray<string>(classNames))) {
```
To avoid to call **ensureIsArray** method for each **bpmnElementId**, I think it's better to call it one time before first **ensureIsArray** calling like done in **addCssClasses(...)**.

tbouffard replied:
Yes, and this will highlight the duplication

csouchet commented on test/unit/component/registry/css-registry.test.ts:
```diff
@@ -13,45 +13,97 @@
+    });
 
-    cssRegistry.addClassNames(bpmnElementId, ['class-name-3', 'class-name-2', 'class-name-4']);
+    it('getClassNames should return a empty array, when no class name is registered for the BPMN element', () => {
```
It should not be in **'Add css classes'** test suite, but in **'Get css classes'** test suite 😉

csouchet commented:
I though we will modify the elements-identification.html page to remove the css class added on the old selected BPMN elements, when we change the selection.

aibcmars reviewed (approved):
LGTM now

tbouffard commented:
@csouchet I have updated the `elements-identification.html` page to remove previous added css classes.
I have also changed the CSS styles: they are now depending on the BPMN kind to highlight found elements differently.

example with `B.2.0`

![elements-detection_remove_and_dedicated_css_classes](https://user-images.githubusercontent.com/27200110/105693149-b158f280-5eff-11eb-8259-4b437086779d.gif)

csouchet commented on test/unit/component/registry/css-registry.test.ts:
```diff
@@ -13,58 +13,118 @@
+      expect(cssRegistry.getClassNames(bpmnElementId)).toHaveLength(0);
+    });
+
+    it('Remove a class that is not present when others have been added first', () => {
```
What happens in this case ?

```javascript
  const bpmnElementId = 'bpmn-id';
  cssRegistry.addClassNames(bpmnElementId, [ 'class-to-remove1', 'class-to-remove2']);

  const result = cssRegistry.removeClassNames(bpmnElementId, ['class-to-remove1', 'class-not-regitered', 'class-to-remove2']);
```

tbouffard replied:
done in 862d84e

tbouffard commented on src/component/registry/css-registry.ts:
```diff
@@ -36,21 +38,24 @@ export class CssRegistry {
-    }
+    const initialClassNames = new Set(this.classNamesByBPMNId.get(bpmnElementId));
+    const resultingClassNames = new Set(initialClassNames);
+    ensureIsArray(classNames).forEach(c => resultingClassNames.add(c));
```
I have just changed this. We only the size of the array prior updating the array and compare the size at the end to know if we added some elements

tbouffard replied:
29db4c9 still update the Set with several add calls. This highlight duplicate/close code between the add and the remove methods.
We may wait for the `toggle` method introduction prior deciding what we do.

csouchet commented on src/component/registry/css-registry.ts:
```diff
@@ -25,7 +26,7 @@ export class CssRegistry {
    */
   getClassNames(bpmnElementId: string): string[] {
-    return this.classNamesByBPMNId.get(bpmnElementId) || [];
+    return Array.from(this.classNamesByBPMNId.get(bpmnElementId) || '');
```
I'm not sure that we want to return an array with an empty string, where there is no registered class, but more a empty array.

tbouffard replied:
This returns an empty array, we have tests that ensure it, see https://github.com/process-analytics/bpmn-visualization-js/blob/29db4c9c94454169d94bd9c226e59a283d96d5be/test/unit/component/registry/css-registry.test.ts#L24-L32
As the syntax seems to be confusing, I can change it (replace the empty string with something different)

tbouffard replied:
Updated in f234b267
